macros/edr/data_monitoring/monitors_query/table_monitoring_query.sql

```diff
@@ -144,4 +91,183 @@
+    ),
+
+    -- divide the freshness metrics above to buckets
+    bucketed_consecutive_updates_freshness as (
+        select
+            edr_bucket_start, edr_bucket_end, update_timestamp, freshness
+        from buckets, consecutive_updates_freshness
+        where update_timestamp >= edr_bucket_start AND update_timestamp < edr_bucket_end
+    ),
+
+    -- we also want to record the freshness at the end of each bucket as an additional point (as it might be larger,
+    -- or even the only point if no data was updated at all in a particular bucket)
+    bucket_end_freshness as (
+        select
+            edr_bucket_start,
+            edr_bucket_end,
+            max(timestamp_val) as update_timestamp,
+            {{ elementary.timediff('second', elementary.cast_as_timestamp('max(timestamp_val)'), "least(edr_bucket_end, {})".format(elementary.current_timestamp_column())) }} as freshness
+        from buckets, unique_timestamps
+        where timestamp_val < edr_bucket_end
+        group by 1,2
+    ),
+
+    -- create a single table with all the freshness values
+    bucket_all_freshness_metrics as (
+        select * from bucketed_consecutive_updates_freshness
+        union all
+        select * from bucket_end_freshness
+    ),
+
+    -- get all the freshness values, ranked by size (we use partition by and not group by, because we also want to have
+    -- the associated timestamp as source value)
+    bucket_freshness_ranked as (
+        select
+            *,
+            rank () over (partition by edr_bucket_end order by freshness desc) as row_number
+        from bucket_all_freshness_metrics
+    )
+
+    select
+        edr_bucket_start,
+        edr_bucket_end,
+        {{ elementary.const_as_string('freshness') }} as metric_name,
+        {{ elementary.cast_as_string('update_timestamp') }} as source_value,
+        freshness as metric_value
+    from bucket_freshness_ranked
+    where row_number = 1
+{% else %}
+    {# Update freshness test not supported when timestamp column is not provided #}
+    {# TODO: We can enhance this test for models to use model_run_results in case a timestamp column is not defined #}
+    {% do return(none) %}
+{% endif %}
+{% endmacro %}
+
+{% macro event_freshness_metric_query(metric_args, timestamp_column) %}
+{% set event_timestamp_column = metric_args.event_timestamp_column %}
+{% set update_timestamp_column = timestamp_column %}
+
+{% if update_timestamp_column %}
+    select
+        edr_bucket_start,
+        edr_bucket_end,
+        {{ elementary.const_as_string('event_freshness') }} as metric_name,
+        {{ elementary.cast_as_string('max({})'.format(event_timestamp_column)) }} as source_value,
+        {{ 'coalesce(max({}), {})'.format(
+                elementary.timediff('second', elementary.cast_as_timestamp(event_timestamp_column), elementary.cast_as_timestamp(update_timestamp_column)),
+                elementary.timediff('second', 'edr_bucket_start', 'edr_bucket_end')
```

Comment:
I was a bit unsure what should I do for the `event_freshness` metric if a particular bucket is empty, since the calculation is local per row.
I thought that maybe saying the freshness is the entire duration of the bucket is good enough (for event data, it's like "infinity"), but maybe there's a better option.